### PR TITLE
Websocket implementation

### PR DESCRIPTION
**Pull request by tommartensen**

### CODE REVIEW

File: src/Utils/ChangeNotifier.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+			this.webSocket.onmessage = this.handleMessage;
+		}
+		this.subscribers = [];
+		this.handleMessage = this.handleMessage.bind(this);
+		this.register = this.register.bind(this);
+		this.unregisterAll = this.unregisterAll.bind(this);
+	}
+	
+	handleNewConnection(event) {
+		console.log("[WEBSOCKET] Established a new connection! Details: "+event.data);
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/ChangeNotifier.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	}
+	
+	handleError(event) {
+		console.log("[WEBSOCKET] An error occured! Details: "+event.data);
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/ChangeNotifier.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+						notificationNote.artifactId === serverNotification.EntityId) || 
+						notificationNote.artifactId === serverNotification.ArtifcatId)) {
+							notificationNote.notificationCallback();
+				}
+			});
+		});
+	}
+
+	handleConnectionClose(event) {
+		console.log("[WEBSOCKET] Connection closed! Details: "+event.data);
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/Utils/ChangeNotifier.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    static isRelevantNotificationForGridView(subscription, notification) {
+        return notification.ArtifactType === "Entity" && subscription.artifactId === notification.ArtifactId;
+    }
+
+	handleNewConnection() {
+		console.log("[WEBSOCKET] Established a new connection! Connection is open.");
+	}
+	
+	handleError() {
+		console.log("[WEBSOCKET] An error occured! Connection could not be established.");
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js:45]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/ChangeNotifier.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+				console.log(subscription);
+				if (ChangeNotifier.isRelevantNotification(subscription, notification)) {
+                    subscription.notificationCallback();
+				}
+			});
+		});
+	}
+
+	handleConnectionClose() {
+		console.log("[WEBSOCKET] Connection closed!");
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js:60]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/ChangeNotifier.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		}
+		this.subscribers = [];
+		this.handleMessage = this.handleMessage.bind(this);
+		this.register = this.register.bind(this);
+		this.unregisterAll = this.unregisterAll.bind(this);
+	}
+
+	static isRelevantNotification(subscription, notification) {
+		return ChangeNotifier.isRelevantNotificationForGridView(subscription, notification);
+        /*return (subscription.artifactType === notification.ArtifactType &&
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this commented out code. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3ACommentedCode)

File: src/Utils/ChangeNotifier.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	
+	handleError() {
+		console.log("[WEBSOCKET] An error occured! Connection could not be established.");
+	}
+	
+	handleMessage(event) {
+		const serverNotifications = JSON.parse(event.data);
+		serverNotifications.forEach((notification) => {
+			this.subscribers.forEach((subscription) => {
+				console.log(subscription);
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/ChangeNotifier.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        ((notification.ArtifactType === "Event" && subscription.artifactId === notification.EntityId)
+        || subscription.artifactId === notification.ArtifactId));*/
+    }
+
+    static isRelevantNotificationForGridView(subscription, notification) {
+        return notification.ArtifactType === "Entity" && subscription.artifactId === notification.ArtifactId;
+    }
+
+	handleNewConnection() {
+		console.log("[WEBSOCKET] Established a new connection! Connection is open.");
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/Utils/ChangeNotifier.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+			(notification.ArtifactType === "Entity" && subscription.artifactId === notification.ArtifactId)
+			||
+			(notification.ArtifactType === "Event" && subscription.artifactId === parseInt(notification.EventTypeId, 10))
+		) {
+			subscription.notificationCallback();
+		}
+	}
+
+	handleNewConnection() {
+		console.log("[WEBSOCKET] Established a new connection! Connection is open.");
```
Comment (hpibptlab) [inline src/Utils/ChangeNotifier.js:41]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/DetailView/DetailView.js [hunk 10]
```diff
@@ -27,6 +27,24 @@ class DetailView extends ConnectionComponent {
...
+		this.registerNotification("Entity", parseInt(this.props.match.params.entityId, 10), this.props.refreshEntity);
+	}
+
+	componentWillUnmount() {
+		this.unregisterAllNotifications();
+	}
+	
+	handleServerSideEventsChanged() {
+		if (this.props.eventTypes.value.length) {
+            this.props.lazyEventLoading(this.state.currentEventType.Id, this.handleEventChange);
```
Comment (MaximilianV) [inline src/DetailView/DetailView.js]: Please use tabs for indentation.

File: src/DetailView/DetailView.js [hunk 11]
```diff
@@ -117,7 +135,7 @@ class DetailView extends ConnectionComponent {
 
 	render() {
 		const entity = this.props.entity.value;
-		const eventTypes = this.props.eventTypes.value;
+		this.eventTypes = this.props.eventTypes.value;
```
Comment (MaximilianV) [inline src/DetailView/DetailView.js]: why was this changed?

File: src/Utils/ConnectionComponent.js [hunk 12]
```diff
@@ -4,9 +4,25 @@ import ErrorMessage from './../Utils/ErrorMessage';
...
 
+	constructor(props) {
+		super(props);
+		if(!window.changeNotifier) {
+			window.changeNotifier = new ChangeNotifier();
+		}
+	}
+	
+	registerNotification(artifactType, artifactId, notificationCallback) {
+		window.changeNotifier.register(artifactType, artifactId, notificationCallback)
```
Comment (MaximilianV) [inline src/Utils/ConnectionComponent.js]: Missing `;` at the EOL

File: src/Utils/ChangeNotifier.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	handleError() {
+		console.log("[WEBSOCKET] An error occured! Connection could not be established.");
+	}
+	
+	handleMessage(event) {
+		const serverNotifications = JSON.parse(event.data);
+		serverNotifications.forEach((notification) => {
+			this.subscribers.forEach((subscription) => {
+				if (ChangeNotifier.handleNotification(subscription, notification)) {
+                    subscription.notificationCallback();
```
Comment (MaximilianV) [inline src/Utils/ChangeNotifier.js]: Please use tabs for indentation.

File: src/Utils/ChangeNotifier.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+				}
+			});
+		});
+	}
+
+	handleConnectionClose() {
```
Comment (MaximilianV) [inline src/Utils/ChangeNotifier.js:59]: What is this function used for (apart from logging)?

**Review by MaximilianV — DISMISSED** [review]

File: src/Utils/ChangeNotifier.js (hunk 14 above)
Comment (tommartensen) [reply src/Utils/ChangeNotifier.js:59]: The connection can also be closed by the server. One could make some error handling at this point.

Comment (MaximilianV) [reply src/Utils/ChangeNotifier.js:59]: Well not that nice, but we can keep it

Comment (tommartensen) [reply src/Utils/ChangeNotifier.js:59]: Maybe when the Snackbar (old SlidingNotification) will be implemented, the closing can be caught here.

File: src/DetailView/DetailView.js (hunk 11 above)
Comment (tommartensen) [reply src/DetailView/DetailView.js]: Was changed during development to fix a bug, is not necessary anymore, so it was changed back. Thanks for noticing!

**Review by frederike-ramin — APPROVED** [review]
